openedx/features/badging/tests/test_helpers.py

```diff
@@ -0,0 +1,215 @@
+    @mock.patch('openedx.features.badging.helpers.filter_earned_badge_by_joined_team')
+    @mock.patch('openedx.features.badging.helpers.is_teams_feature_enabled')
+    @mock.patch('openedx.features.badging.helpers.get_course_by_id')
+    def test_course_badges_user_not_joined_any_course_team(self, mock_get_course_by_id,
+                                                           mock_is_teams_feature_enabled,
+                                                           mock_filter_earned_badge_by_joined_team,
+                                                           mock_add_badge_earned_date):
+        mock_get_course_by_id.return_value = mock.ANY
+        mock_is_teams_feature_enabled.return_value = True
+        mock_filter_earned_badge_by_joined_team.return_value = False, list()
+
+        badge = BadgeFactory(type=self.type_conversationalist, threshold=2)
+
+        badges = badge_helpers.get_course_badges(self.user, self.course1.id, earned_badges=mock.ANY)
+
+        expected_result = {
+            'badges': {
+                'conversationalist': [{
+                    'description': None, 'image': badge.image, 'threshold': 2L,
+                    'date_created': badge.date_created, 'type': u'conversationalist',
+                    u'id': badge.id, 'name': badge.name
+                }],
+                'team': []
+            },
+            'team_joined': False
+        }
+
+        self.assertEqual(expected_result, badges)
+
+    def test_badge_earned_date_no_course_badges(self):
+        with self.assertRaises(TypeError):
+            badge_helpers.add_badge_earned_date(self.course1.id, course_badges=None, earned_badges=mock.ANY)
+
+    def test_badge_earned_date_no_earned_badges(self):
+        badge = BadgeFactory(type=self.type_conversationalist, threshold=2)
+        course_badges = [badge, ]
+        with self.assertRaises(TypeError):
+            badge_helpers.add_badge_earned_date(self.course1.id, course_badges, earned_badges=None)
+
+    def test_badge_earned_date_course_badges_subset_of_earned_badges(self):
+        course2 = CourseFactory(org="test2", number="123", run="1", display_name="XYZ")
+
+        badge1 = BadgeFactory(type=self.type_conversationalist, threshold=2)
+        badge2 = BadgeFactory(type=self.type_conversationalist, threshold=5)
+
+        user_badge_1 = UserBadgeFactory(user=self.user, course_id=self.course1.id, badge=badge1)
+        user_badge_2 = UserBadgeFactory(user=self.user, course_id=course2.id, badge=badge2)
+
+        course_badges = list(
+            Badge.objects.all().values()
+        )
+        earned_badges = [user_badge_1, user_badge_2]
+        badge_helpers.add_badge_earned_date(self.course1.id, course_badges, earned_badges)
+
+        # get badges by index, since list maintains insertion order
+        self.assertIsNotNone(course_badges[0]['date_earned'])
+        self.assertRaises(KeyError, lambda: course_badges[1]['date_earned'])
```

Comment:
let's verify the result of this test

Reply:
It is passing, I have also added documentation